syncer/syncer.go

```diff
@@ -0,0 +1,272 @@
+
+	go s.initializePeerMap()
+	go s.startPeerConnectionEventProcess()
+
+	return nil
+}
+
+func (s *syncer) initializePeerMap() {
+	peerStatuses := s.syncPeerClient.GetConnectedPeerStatuses()
+	s.peerMap.PutPeers(peerStatuses)
+
+	for peerStatus := range s.syncPeerClient.GetPeerStatusUpdateCh() {
+		s.peerMap.Put(peerStatus)
+
+		select {
+		case s.newStatusCh <- struct{}{}:
+		default:
+		}
+	}
+}
+
+func (s *syncer) startPeerConnectionEventProcess() {
+	for e := range s.syncPeerClient.GetPeerConnectionUpdateEventCh() {
+		peerID := e.PeerID
+
+		switch e.Type {
+		case event.PeerConnected:
+			go func() {
+				status, err := s.syncPeerClient.GetPeerStatus(peerID)
+				if err != nil {
+					s.logger.Warn("failed to get peer status, skip", "id", peerID, "err", err)
+
+					return
+				}
+
+				s.peerMap.Put(status)
+
+				select {
+				case s.newStatusCh <- struct{}{}:
+				default:
+				}
+			}()
+		case event.PeerDisconnected:
+			s.peerMap.Remove(peerID)
+		}
+	}
+}
+
+func (s *syncer) GetSyncProgression() *progress.Progression {
+	return s.syncProgression.GetProgression()
+}
+
+// HasSyncPeer returns whether syncer has the peer to syncs blocks
+// return false if syncer has no peer whose latest block height doesn't exceed local height
+func (s *syncer) HasSyncPeer() bool {
+	if s.peerMap == nil {
+		return false
+	}
+
+	bestPeer := s.peerMap.BestPeer(nil)
+	header := s.blockchain.Header()
+
+	return bestPeer != nil && bestPeer.Number > header.Number
+}
+
+func (s *syncer) BulkSync(ctx context.Context, newBlockCallback func(*types.Block) bool) error {
+	localLatest := uint64(0)
+	updateLocalLatest := func() {
+		if header := s.blockchain.Header(); header != nil {
+			localLatest = header.Number
+		}
+	}
+
+	updateLocalLatest()
+
+	// Create a blockchain subscription for the sync progression and start tracking
+	s.syncProgression.StartProgression(localLatest+1, s.blockchain.SubscribeEvents())
+
+	// Stop monitoring the sync progression upon exit
+	defer s.syncProgression.StopProgression()
+
+	skipList := make(map[peer.ID]bool)
+
+	for {
+		bestPeer := s.peerMap.BestPeer(skipList)
+		if bestPeer == nil || bestPeer.Number <= localLatest {
+			break
+		}
+
+		// Set the target height
+		s.syncProgression.UpdateHighestProgression(bestPeer.Number)
+
+		lastNumber, _, err := s.bulkSyncWithPeer(bestPeer.ID, newBlockCallback)
+		if err != nil {
+			s.logger.Warn("failed to complete bulk sync with peer, try to next one", "peer ID", bestPeer.ID)
+		}
+
+		// if node could sync with the peer fully, then exit loop
+		if err == nil && lastNumber >= bestPeer.Number {
+			break
+		}
+
+		updateLocalLatest()
+
+		skipList[bestPeer.ID] = true
+	}
+
+	return nil
+}
+
+func (s *syncer) WatchSync(ctx context.Context, callback func(*types.Block) bool) error {
+	localLatest := s.blockchain.Header().Number
+	skipList := make(map[peer.ID]bool)
+
+	// Loop until context is canceled
+	for {
+		//Wait for a new event to arrive
+		<-s.newStatusCh
```

Comment:
Fixed at d75dc06974a15bbf55ba2bc69f8173b95e531db2

Reply:
Where do you call `Syncer.Close()`?

Reply:
from ibft, you can check in latest code

Reply:
@zivkovicmilos It will be called if when we are closing ibft consensus https://github.com/0xPolygon/polygon-edge/blob/feature/sync-v2/consensus/ibft/ibft.go#L1375